### PR TITLE
Parallelize building and pushing Docker images

### PR DESCRIPTION
**Pull request by Mandrenkov**
This PR contributes two new Actions:
* **Build Docker Images**: Extension of `build-docker-image` which builds Docker images in parallel.
* **Push Docker Images**: Extension of `push-docker-image` which pushes Docker images in parallel.

### CODE REVIEW

**Review by anthayes92 — APPROVED** [review]
Nice parallelisation! :zap: :construction_worker_man:  Only question is how did you test this works?

File: build-docker-images/action.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+            docker build \
+              --build-arg COMMIT_SHA=${{ github.sha }} \
+              --ssh default \
+              --tag $target \
+              --target $target \
+              ${{ inputs.docker-build-extra-args }} \
+              . &
+        done
+
+        wait
```
Comment (michaelvu97) [inline build-docker-images/action.yml:33]: 🧠 clever

File: push-docker-images/action.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        fi
+
+        for i in "${!images[@]}"; do
+          for tag in "${tags[@]}"; do
+            image="${images[$i]}"
+            repo="${repos[$i]}"
+            target="${repo}:${tag}"
+
+            docker tag $image $target
+            docker push $target &
```
Comment (michaelvu97) [inline push-docker-images/action.yml:40]: If there eare e.g. 3 tags and 3 images, does this `docker push` execute 9 times? I might be wrong, but wouldn't it suffice to `tag` the image `T` times, then `push` once?

Or is it required to `docker push` every `repo:tag` pair?

**Review by michaelvu97 — APPROVED** [review]
☁️ Nicely done

File: push-docker-images/action.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+inputs:
+  images:
+    description: >
+      Docker images to push (local tags or IDs) as a space-separated list.
+      The number of images must match the number of repositories.
+    required: true
+  repos:
+    description: >
+      Docker image repositories (to push the images to) as a space-separated list.
+      The number of repositories must match the number of images.
```
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply push-docker-images/action.yml:13]: Not directly but we can include a check in a `step` as done below!

File: push-docker-images/action.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  using: composite
+  steps:
+    - shell: bash
+      run: |
+        read -ra images <<< "${{ inputs.images }}"
+        read -ra repos <<< "${{ inputs.repos }}"
+        IFS="," read -ra tags <<< "${{ inputs.tags }}"
+
+        if [ "${#images[@]}" -ne "${#repos[@]}" ]; then
+          echo "Number of images must match the number of repositories."
```
[the remark this answers is not in the document]
Comment (Mandrenkov) [reply push-docker-images/action.yml:29]: Yep, exactly!

File: build-docker-images/action.yml (hunk 1 above)
Comment (Mandrenkov) [reply build-docker-images/action.yml:33]: Thanks! Something something shoulders of giants.

**Comment by Mandrenkov** [comment]
> Nice parallelisation! ⚡ 👷‍♂️ Only question is how did you test this works?

Thanks! I tested these changes by running the Actions from a Workflow in another repository. You should be tagged as a reviewer on that other PR too!

File: push-docker-images/action.yml [hunk 5]
```diff
@@ -31,14 +31,14 @@ runs:
...
-          for tag in "${tags[@]}"; do
-            image="${images[$i]}"
-            repo="${repos[$i]}"
-            target="${repo}:${tag}"
+          image="${images[$i]}"
+          repo="${repos[$i]}"
 
-            docker tag $image $target
-            docker push $target &
+          for tag in "${tags[@]}"; do
```
Comment (michaelvu97) [inline push-docker-images/action.yml:37]: 🔥

**Pull request merged**